presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EliminateCrossJoins.java

```diff
@@ -50,7 +50,7 @@
 public class EliminateCrossJoins
         implements Rule
 {
-    private static final Pattern PATTERN = Pattern.node(JoinNode.class);
+    private static final Pattern PATTERN = Pattern.matchByClass(JoinNode.class);
```

Comment:
Great idea, as in other such things lets make it a followup step. I do not want to make this PR to grow and grow, but rather to get it merged and to solve issue you suggested.